src/latex2sympy/latex2antlrJson.cpp

```diff
@@ -0,0 +1,294 @@
+                Json::Value childNode = toJson(child, parser);
+                if (node[childName].empty()) {
+                    // force certain rules to ALWAYs return as an array
+                    // otherwise, single values are set as property objects, e.g. { "foo": child }
+                    if (childName == "postfix" ||
+                        childName == "postfix_nofunc" ||
+                        childName == "postfix_op" ||
+                        childName == "matrix_row")
+                    {
+                        Json::Value array;
+                        array.append(childNode);
+                        node[childName] = array;
+                    } else {
+                        node[childName] = childNode;
+                    }
+                } else if (node[childName].isArray()) {
+                    // if there is already an array at the given name, simply append the new child
+                    node[childName].append(childNode);
+                } else {
+                    // if a child with the same name was added as a property object
+                    // convert that property to an array, and append the new child
+                    // e.g. { "foo": child1 } => { "foo": [child1, child2]}
+                    Json::Value array;
+                    array.append(node[childName]);
+                    array.append(childNode);
+                    node[childName] = array;
+                }
+            }
+        } else if (tree::TerminalNode::is(child)) {
+            tree::TerminalNode *terminalNode = static_cast<tree::TerminalNode *>(child);
+            Json::Value childNode = toJson(terminalNode);
+            tokens.append(childNode);
+        }
+    }
+
+    // if there is only one token, merge it into the current node
+    // otherwise add tokens as a nested array
+    if (tokens.size() == 1) {
+        node["text"] = tokens[0]["text"];
+        node["type"] = tokens[0]["type"];
+    } else if (tokens.size() > 0) {
+        node["tokens"] = tokens;
+    }
+
+    // return the full text of the tree for specific rules
+    if (name == "func_multi_arg" ||
+        name == "mathit_text" ||
+        parentName == "supexpr" ||
+        parentName == "subexpr" ||
+        parentName == "accent") {
+        node["text"] = tree->getText();
+    }
+
+    return node;
+}
+
+/// @brief Convert a ParseTree to JSON and then write it to a string.
+std::string toJsonString(tree::ParseTree *tree, LATEXParser *parser) {
+    Json::Value root = toJson(tree, parser);
+    Json::FastWriter fastWriter;
+    std::string json = fastWriter.write(root);
+    return json;
+}
+
+class MathErrorListener : public BaseErrorListener {
+    public:
+        std::string src;
+        MathErrorListener(const std::string &input) : BaseErrorListener() {
+            src = input;
+        }
+
+        void syntaxError(Recognizer *recognizer, Token *offendingSymbol, size_t line, size_t charPositionInLine,
```

Comment:
my gut says there is a prettier way to build these error strings, but I don't have a suggestion at this point.

Reply:
@lily514 yeah i was mostly attempting to match how it used to be in python, i feel it got fairly close

but yeah definitely had to look up literally every c++ thing in here

Reply:
Ope, here it is, you can do use standard format to build a new string directly, instead of doing the string stream.

str = std::format("{1}\n{2}\n{3}^", msg, src, marker);

Reply:
actually you don't need the numbers in there as long as you put in the arguments in order
str = std::format("{}\n{}\n{}^", msg, src, marker);
https://en.cppreference.com/w/cpp/utility/format/format

Reply:
I did try that initially but it isn't available

![image](https://user-images.githubusercontent.com/689209/217634658-734bf8db-c42f-4ad0-b74f-49e69f6574f1.png)

Reply:
oh i might need an include

Reply:
:no_idea_dog:

Reply:
ah well, like you said what you have is working. I think I'm ready to approve it at this point.